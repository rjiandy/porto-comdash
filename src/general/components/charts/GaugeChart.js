// @flow

import React from 'react';
import {Cell, PieChart, Pie, ResponsiveContainer, Tooltip} from 'recharts';
import {StyleSheet} from 'react-primitives';

import {ANIMATION_DURATION} from '../../constants/chartSettings';
import {View, Text} from '../../components/coreUIComponents';
import {WHITE, GREY, GAUGE_CHART_COLOR} from '../../constants/colors';
import {SMALL_FONT_SIZE} from '../../constants/text';
import roundDecimal from '../../helpers/roundDecimal';

type Shape = {
  cx: number;
  cy: number;
  outerRadius: number;
  midAngle: number;
};

type Props = {
  value: number;
  width: number;
  toolTipLabel: string;
  height?: number;
  colorData?: Array<{value: number; color: string}>;
  offset?: number;
};

const DEFAULT_COLOR_DATA = [
  {
    value: 25,
    color: GAUGE_CHART_COLOR.red,
  },
  {
    value: 25,
    color: GAUGE_CHART_COLOR.yellow,
  },
  {
    value: 25,
    color: GAUGE_CHART_COLOR.yellow,
  },
  {
    value: 25,
    color: GAUGE_CHART_COLOR.green,
  },
];

export default function GaugeChart(props: Props) {
  let {height, value: chartValue, width, toolTipLabel, offset} = props;
  let colorData = props.colorData || DEFAULT_COLOR_DATA;
  chartValue = offset ? chartValue - offset : chartValue;

  const activeSectorIndex = colorData
    .map((cur, index, arr) => {
      const curMax = [...arr]
        .splice(0, index + 1)
        .reduce((a, b) => ({value: a.value + b.value})).value;
      return chartValue > curMax - cur.value && chartValue <= curMax;
    })
    .findIndex((cur) => cur);

  const sumValues = colorData.reduce((a, b) => a + b.value, 0);

  const arrowData = [
    {value: chartValue},
    {value: 0},
    {value: sumValues - chartValue},
  ];

  const pieProps = {
    startAngle: 180,
    endAngle: 0,
    cx: width / 2,
    cy: width / 2,
  };

  const pieRadius = {
    outerRadius: width / 2 * 0.7,
  };

  const Arrow = ({cx, cy, midAngle, outerRadius}: Shape) => {
    //eslint-disable-line react/no-multi-comp
    const RADIAN = Math.PI / 180;
    const sin = Math.sin(-RADIAN * midAngle);
    const cos = Math.cos(-RADIAN * midAngle);
    const mx = cx + (outerRadius + width * 0.03) * cos;
    const my = cy + (outerRadius + width * 0.03) * sin;
    return (
      <g>
        <circle
          cx={cx}
          cy={cy}
          r={width * 0.02}
          fill={GAUGE_CHART_COLOR.needle}
          stroke="none"
        />
        <path
          d={`M${cx},${cy}L${mx},${my}`}
          strokeWidth="5"
          stroke={GAUGE_CHART_COLOR.needle}
          fill="none"
          strokeLinecap="round"
        />
      </g>
    );
  };

  return (
    <ResponsiveContainer width={width} height={height || '100%'}>
      <PieChart>
        <Pie
          activeIndex={activeSectorIndex}
          data={colorData}
          animationDuration={ANIMATION_DURATION}
          label={(props: Object) => {
            let {index, value, x, y} = props;
            let offsetX = 0;
            let offsetY = 0;
            if (x <= 50) {
              offsetX = index * -15;
              offsetY = 25 + index * 6;
            } else if (x <= 100) {
              offsetX = index * -15;
              offsetY = 25;
            } else if (x <= 150) {
              offsetX = index * -14;
              offsetY = 15;
            } else if (x <= 180) {
              offsetX = index * -11;
            } else {
              offsetX = index * -7;
              offsetY = -10;
            }
            return (
              <g>
                <text
                  x={x + 5 + offsetX}
                  y={y + offsetY}
                  fontSize={SMALL_FONT_SIZE}
                >
                  {offset + index * value}
                </text>
                {index === colorData.length - 1 && (
                  <text
                    x={x + 7 - index * 5}
                    y={y + 23}
                    fontSize={SMALL_FONT_SIZE}
                  >
                    {offset + (index + 1) * value}
                  </text>
                )}
              </g>
            );
          }}
          labelLine={false}
          {...pieRadius}
          {...pieProps}
        >
          {colorData.map((entry, index) => (
            <Cell key={`cell-${index}`} fill={colorData[index].color} />
          ))}
        </Pie>
        <Pie
          stroke="none"
          activeIndex={1}
          activeShape={Arrow}
          data={arrowData}
          fill="none"
          {...pieProps}
        />
        <Tooltip
          content={(args: {active: boolean}) => (
            <GaugeChartTooltip
              active={args.active}
              value={chartValue + offset}
              label={toolTipLabel}
            />
          )}
        />
      </PieChart>
    </ResponsiveContainer>
  );
}

type ToolTipProps = {
  label: string;
  active: boolean;
  value: number;
};

function GaugeChartTooltip(props: ToolTipProps) {
  const {active, value, label} = props;

  if (active) {
    return (
      <View style={styles.tooltipContainer}>
        <Text>{`${label} : ${roundDecimal(value)}%`}</Text>
      </View>
    );
  }

  return null;
}

const styles = StyleSheet.create({
  tooltipContainer: {
    backgroundColor: WHITE,
    borderColor: GREY,
    borderWidth: 1,
    padding: 10,
  },
});
