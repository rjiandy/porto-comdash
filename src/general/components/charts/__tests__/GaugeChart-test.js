// @flow

import React from 'react';
import {shallow} from 'enzyme';
import toJSON from 'enzyme-to-json';

import GaugeChart from '../GaugeChart';

it('Should render gauge bar chart correctly', () => {
  let gaugeChart = shallow(
    <GaugeChart value={10.5} width={100} toolTipLabel="Test Tooltip" />
  );
  expect(toJSON(gaugeChart)).toMatchSnapshot();
});
