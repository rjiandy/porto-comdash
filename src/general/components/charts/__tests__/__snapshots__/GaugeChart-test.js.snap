// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render gauge bar chart correctly 1`] = `
<ResponsiveContainer
  debounce={0}
  height="100%"
  width={100}
>
  <PieChart
    barCategoryGap="10%"
    barGap={4}
    cx="50%"
    cy="50%"
    endAngle={360}
    innerRadius={0}
    layout="centric"
    margin={
      Object {
        "bottom": 5,
        "left": 5,
        "right": 5,
        "top": 5,
      }
    }
    outerRadius="80%"
    stackOffset="none"
    startAngle={0}
  >
    <Pie
      activeIndex={0}
      animationBegin={400}
      animationDuration={500}
      animationEasing="ease"
      cx={50}
      cy={50}
      data={
        Array [
          Object {
            "color": "#FF1E3C",
            "value": 25,
          },
          Object {
            "color": "#FFD700",
            "value": 25,
          },
          Object {
            "color": "#FFD700",
            "value": 25,
          },
          Object {
            "color": "#63A500",
            "value": 25,
          },
        ]
      }
      endAngle={0}
      fill="#808080"
      hide={false}
      innerRadius={0}
      isAnimationActive={true}
      label={[Function]}
      labelLine={false}
      legendType="rect"
      minAngle={0}
      outerRadius={35}
      paddingAngle={0}
      startAngle={180}
      stroke="#fff"
    >
      <Cell
        fill="#FF1E3C"
      />
      <Cell
        fill="#FFD700"
      />
      <Cell
        fill="#FFD700"
      />
      <Cell
        fill="#63A500"
      />
    </Pie>
    <Pie
      activeIndex={1}
      activeShape={[Function]}
      animationBegin={400}
      animationDuration={1500}
      animationEasing="ease"
      cx={50}
      cy={50}
      data={
        Array [
          Object {
            "value": 10.5,
          },
          Object {
            "value": 0,
          },
          Object {
            "value": 89.5,
          },
        ]
      }
      endAngle={0}
      fill="none"
      hide={false}
      innerRadius={0}
      isAnimationActive={true}
      labelLine={true}
      legendType="rect"
      minAngle={0}
      outerRadius="80%"
      paddingAngle={0}
      startAngle={180}
      stroke="none"
    />
    <Tooltip
      active={false}
      animationDuration={400}
      animationEasing="ease"
      content={[Function]}
      coordinate={
        Object {
          "x": 0,
          "y": 0,
        }
      }
      cursor={true}
      cursorStyle={Object {}}
      isAnimationActive={true}
      itemSorter={[Function]}
      itemStyle={Object {}}
      labelStyle={Object {}}
      offset={10}
      separator=" : "
      viewBox={
        Object {
          "x1": 0,
          "x2": 0,
          "y1": 0,
          "y2": 0,
        }
      }
      wrapperStyle={Object {}}
    />
  </PieChart>
</ResponsiveContainer>
`;
